include/igl/dual_contouring.cpp

```diff
@@ -0,0 +1,559 @@
+#include "dual_contouring.h"
+#include "quadprog.h"
+#include "parallel_for.h"
+#include <thread>
+#include <mutex>
+#include <vector>
+#include <unordered_map>
+#include <iostream>
+
+namespace igl
+{
+  // These classes not intended to be used directly
+  class Hash
+  {
+    public:
+      // https://stackoverflow.com/a/26348708/148668
+      uint64_t operator()(const std::tuple<int, int, int> & key) const
+      {
+        // Check that conversion is safe. Could use int16_t directly everywhere
+        // below but it's an uncommon type to expose and grid indices should
+        // never be more than 2¹⁵-1 in the first place.
+        assert( std::get<0>(key) == (int)(int16_t)std::get<0>(key));
+        assert( std::get<1>(key) == (int)(int16_t)std::get<1>(key));
+        assert( std::get<2>(key) == (int)(int16_t)std::get<2>(key));
+        uint64_t result = uint16_t(std::get<0>(key));
+        result = (result << 16) + uint16_t(std::get<1>(key));
+        result = (result << 16) + uint16_t(std::get<2>(key));
+        return result;
+      };
+  };
+  template <typename Scalar>
+  class DualContouring
+  {
+    // Types
+    public:
+      typedef Eigen::Matrix<Scalar,1,3> RowVector3S;
```

Comment:
Small comment, but I find the syntax `using Foo = Bar;` more clear for typedefs (plus it also works with templated typedefs).